docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,311 @@
+* **Claim**: A resource that represents a claim on an IP address.
+* **Load Balancer Controller**: a controller capable of fulfilling all Ingress paths for a given class of claims.
+
+__Example__:
+* Cluster is bootstrapped with kube-controller-manager --loadbalancer-controllers=(cloud=gce),haproxy
+* User creates svc1 and svc2
+* User creates claim1 to get a public ip.
+  GCEController:
+    - receives watch notification for `Class: cloud` object.
+    - allocates cloud resources for a new L7
+    - spins off a goroutine that watches services for that one claim and updates the L7
+* User creates claim2.
+  Haproxy Controller:
+    - receives watch notification for `Class: haproxy` object
+    - creates a haproxy pod
+    - haproxy pod watches services for that one claim and reconfigures haproxy
+* We end up with 2 ips:
+    - 172.12.14.111:80 serving foo and bar at L7 through a GCE L7
+    - 172.11.14.143:8080 switching between foobar and baz with different certs using sni at L4, through a pod in the cluster.
+
+```
+                Cloud                                  HaProxy pod
+                  ^                                         ^
+                  |                                         |
+            --------------------                     -------------------------
+           | GCE Controller pod |                   | Ha proxy Controller pod |
+            --------------------                     -------------------------
+                  |                                         |
+                  v                                         v
+            {Name: claim1,                          {Name: claim2,
+             Class: cloud,                           Class: bare-metal,
+             Ip: 172.11.14.111}                      Ip: 172.11.14.143:8080}
+                 /       \                              /        \
+             Path1      Path2                       Path3        Path4
+             type: l7   type: l7                    type: l4     type: l4
+             foo.com    bar.com                     foobar.com   baz.com
+             /a - svc1  /prod - svc2                cert: c1     cert: c2
+             /b - svc1                              svc4         svc5
+```
+
+#### Ingress Path
+
+This is an example ingress path. It encapsulates:
+* L7 proxying information (i.e map these urls to these backends)
+* Security configuration (i.e tlsMode, secret)
+* Metadata needed for more advanced loadbalancing (e.g session persistence)
+
+See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
+
+```yaml
+apiVersion: v1
+kind: IngressPath
+metadata:
+  name: l7ingress
+  type: l7
+spec:
+  host: www.example.com
+  tlsMode: Termination
+  secret: foosecret
+  pathMap:
```

Comment:
Aren't maps verboten?  Also, what if I don't want to allow users to pick a
domain, but give them one?  Do they just fill out random values?  Array
fixes both of those problems.

On Aug 18, 2015, at 8:47 PM, Prashanth B <notifications@github.com> wrote:

In docs/proposals/loadbalancing.md
<https://github.com/kubernetes/kubernetes/pull/12827#discussion_r37367090>:

> +* Security configuration (i.e tlsMode, secret)
> +* Metadata needed for more advanced loadbalancing (e.g session persistence)
> +
> +See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
> +
> +```yaml
> +apiVersion: v1
> +kind: IngressPath
> +metadata:
> +  name: l7ingress
> +  type: l7
> +spec:
> +  host: www.example.com
> +  tlsMode: Termination
> +  secret: foosecret
> +  pathMap:

@brendandburns <https://github.com/brendandburns> suggested including
subdomain in a single resource (something I'd planned to achieve via
joining). His reason makes sense, from a UX perspective having a single
place to manage the ingress points of an entire website is better.

I experimented a little with:
flat list

  pathMap:
  "foo.domain.tld/foo/*"
    service:
    port:
  "foo.domain.tld/bar/*"
    service:
    port:
  "bar.domain.tld/foobar/*"
    service:
    port:

vs nested path matchers

  pathMap:
  "foo.domain.tld"
    "/foo/*":
      service:
      port:
    "/bar/*":
      service:
      port:
  "bar.domain.tld"
    "/foobar/*":
      service:
      port:

I find the latter easier to think about. So I'm goint to convert this from
the map of path->backend that it currently is, to a nested map. In the
simple case this nested map will only have one subdomain (*). The host in
the spec will be used to validate that there's only a single domain per
ingress point.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12827/files#r37367090>.